devlib/instrument/baylibre_acme.py

```diff
@@ -0,0 +1,424 @@
+                      for d, name in zip(self.iio_context.devices, probe_names)]
+        else:
+            probes = [IIOINA226Instrument(d) for d in self.iio_context.devices]
+
+        self.probes = {p.nickname : p for p in probes}
+
+    # channel management
+
+    def list_channels(self):
+        # TODO: not sure what to do:
+        pass
+        #      or
+        # return {p: self.probes[p].list_channels() for p in self.probes}
+
+    @staticmethod
+    def add_channel():
+        pass # Channels are handled by the probes
+
+    # properties
+
+    sample_rate_hz = _BaylibreAcmePropertyFromProbes('sample_rate_hz', True)
+    oversampling_ratios_available = \
+            _BaylibreAcmePropertyFromProbes('oversampling_ratios_available')
+    integration_times_available = \
+            _BaylibreAcmePropertyFromProbes('integration_times_available')
+
+    # initialization and teardown
+
+    def setup(self, shunt_resistor,
+              integration_time_bus,
+              integration_time_shunt,
+              oversampling_ratio,
+              buffer_samples_count=None,
+              buffer_is_circular=False,
+              absolute_timestamps=False):
+
+        args = {k : v for k, v in locals().iteritems() if k != 'self'}
+
+        def pseudo_list(v, i):
+            try:
+                return v[i]
+            except TypeError:
+                return v
+
+        for i, p in enumerate(self.probes.values()):
+            p.setup(**{k : pseudo_list(v, i) for k, v in args.iteritems()})
+
+    def reset(self, probes=None, sites=None, kinds=None, channels=None):
+        if isinstance(probes, basestring):
+            probes = [probes]
+
+        self.active_probes = [] # List of the names of the active probes
+        for p in (probes or self.probes):
+            try:
+                self.probes[p].reset(sites, kinds, channels)
+                self.active_probes.append(p)
+            except KeyError as e:
+                self.teardown()
+                raise KeyError('Unknown probe {}'.format(e))
+            except ValueError:
+                self.teardown() # undo what was done by this function
+                raise
+
+    def teardown(self):
+        for p in self.active_probes:
+            self.probes[p].teardown()
+        del self.active_probes[:]  # Py3: self.active_probes.clear()
+
+    def start(self):
+        for p in self.active_probes:
+            self.probes[p].start()
+
+    def stop(self):
+        for p in self.active_probes:
+            self.probes[p].stop()
+
+    def get_data(self, outfile=None):
+        def compose_fname(basefile, probe_name):
+            if basefile:
+                base, ext = os.path.splitext(basefile)
+                return '{}_{}{}'.format(base, probe_name, ext)
+            return basefile
+
+        pairs = ((p, self.probes[p].get_data(compose_fname(outfile, p)))
+                 for p in self.active_probes)
+        return collections.OrderedDict(pairs)
```

Comment:
With multiple active probes, we are returning a dictionary of `MeasurementCsv`s which does not comply with the parent interface!

I thought about merging all the probes into a single CSV but:

- as probes can have different sampling rates, the CSV for a given probe may have more samples (*i.e.* more lines) than the CSV for another probe for a single run. A solution would be to prevent the user from having different sampling rates but isn't this a bit extreme?

- Each probe reports its own timestamps so that samples from different probes (even if they are taken at the same sampling frequency) are not taken at the same times which would mean that a CSV for multiple probes would contain multiple timestamp columns!

Reply:
Align the timestamps and re-sample to the highest rate the individual outputs before combining them. This ought to be relatively straight-forward to do with `pandas`.